genesis/streaming.py

```diff
@@ -14,350 +14,414 @@
+                    msg.value(),
+                    Metadata(
+                        msg.topic(),
+                        msg.partition(),
+                        msg.offset(),
+                        msg.timestamp()[1],
+                        msg.key(),
+                        self._idx,
+                    ),
+                )
+                self._idx += 1
+            elif msg.error().code() == KafkaError._PARTITION_EOF:
+                # silently skip _PARTITION_EOF messages
+                continue
+            else:
+                raise Exception(msg.error())
+
+    def _raw_stream(self, timeout):
+        last_msg_time = time.time()
+        while True:
+            msgs = self.c.consume(1000, 1.0)
+            if not msgs:
+                if timeout and (time.time() - last_msg_time >= timeout):
+                    return
+                yield HEARTBEAT_SENTINEL
+            else:
+                # unpack messages and check for errors
+                yield list(self._validate_and_process(msgs))
+
+                # reset the timeout counter
+                last_msg_time = time.time()
+
+            # actualy commit any offsets the user committed
+            self._commit_to_kafka(defer=True)
+
+    # returns a generator returning deserialized, user-processed messages
+    def _filtered_stream(self, mapper, filter, timeout):
+        t_last = time.time()
+
+        for msgs in itertools.chain(
+            [self._buffer.values()], self._raw_stream(timeout=timeout)
+        ):
+            if not is_heartbeat(msgs):
+                self._buffer = dict(enumerate(msgs))
+
+                # process the messages on the workers
+                for i, (rec, meta) in enumerate(
+                    mapper(ParseAndFilter(parser=self._parser, filter=filter), msgs)
+                ):
+                    # pop the message from the buffer, indicating we've processed it
+                    del self._buffer[i]
+
+                    # mark as consumed and increment the index _before_ we yield,
+                    # as we may never come back from yielding (if the user decides
+                    # to break from the loop)
+                    self._consumed[(meta.topic, meta.partition)] = meta.offset
+
+                    # yield if the filter didn't return None
+                    if rec is not None:
+                        yield (rec, meta)
+
+    def commit(self, defer=True):
+        self._committed = self._consumed.copy()
+        self._commit_to_kafka(defer=defer)
+
+    def _commit_to_kafka(self, defer=True):
+        # Occasionally commit read offsets (note: while this ensures we won't lose any
+        # messages, some may be duplicated if the program is killed before offsets are committed).
+        now = time.time()
+        last = self.last_commit_time if self.last_commit_time is not None else 0
+        if self._committed and (not defer or now - last > 5.0):
+            print("COMMITTING", file=sys.stderr)
```

Comment:
Ugh, this is a leftover debug statement. As you continue to clean up this code, things like these should go to the debug log via the logging module.